relay-event-normalization/src/normalize/span/tag_extraction.rs

```diff
@@ -354,11 +355,60 @@ static SQL_TABLE_EXTRACTOR_REGEX: Lazy<Regex> = Lazy::new(|| {
     Regex::new(r#"(?i)(from|into|update)(\s|")+(?P<table>(\w+(\.\w+)*))(\s|")+"#).unwrap()
 });
 
-/// Returns the table in the SQL query, if any.
-///
-/// If multiple tables exist, only the first one is returned.
-fn sql_table_from_query(query: &str) -> Option<&str> {
-    extract_captured_substring(query, &SQL_TABLE_EXTRACTOR_REGEX)
+/// Used to replace placeholders (parameters) by dummy values such that the query can be parsed.
+static SQL_PLACEHOLDER_REGEX: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"(?:\?+|\$\d+|%(?:\(\w+\))?s|:\w+)").unwrap());
+
+/// Returns a sorted, comma-separated list of SQL tables, if any.
+fn sql_tables_from_query(db_system: Option<&str>, query: &str) -> Option<String> {
+    use sqlparser::ast::{ObjectName, Visit, Visitor};
+    use sqlparser::dialect::GenericDialect;
+    use std::ops::ControlFlow;
+    /// Visitor that finds relation names.
+    struct RelationsVisitor {
+        relations: BTreeSet<String>,
+    }
+
+    impl Visitor for RelationsVisitor {
+        type Break = ();
+
+        fn pre_visit_relation(&mut self, relation: &ObjectName) -> ControlFlow<Self::Break> {
+            if let Some(name) = relation.0.last() {
+                let last = name.value.split('.').last().unwrap_or(&name.value);
+                self.relations.insert(last.to_lowercase());
+            }
+            ControlFlow::Continue(())
+        }
+    }
+
+    // See https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md#notes-and-well-known-identifiers-for-dbsystem
+    //     https://docs.rs/sqlparser/latest/sqlparser/dialect/fn.dialect_from_str.html
+    let dialect = db_system
+        .and_then(sqlparser::dialect::dialect_from_str)
+        .unwrap_or_else(|| Box::new(GenericDialect {}));
+
+    let parsable_query = SQL_PLACEHOLDER_REGEX.replace_all(query, "1");
+    match sqlparser::parser::Parser::parse_sql(&*dialect, &parsable_query) {
+        Ok(ast) => {
+            let mut visitor = RelationsVisitor {
+                relations: Default::default(),
+            };
+            ast.visit(&mut visitor);
```

Comment:
Does the order of the tables in the output depend on the implementation details of `visit`? Should we sort the table names before adding them to a string? Should we even care about it?

Reply:
The order is maintained by the `BTreeSet` (added a comment to clarify). We definitely care because having two different tag values containing the same tables produces needless cardinality.

Reply:
Oh I missed that! Good stuff.